exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -379,11 +379,16 @@ public FragmentExecutor getFragmentRunner(final FragmentHandle handle) {
       return runningFragments.get(handle);
     }
 
+    /**
+     * receive the RuntimeFilter thorough the wire
+     * @param runtimeFilter
+     */
     public void receiveRuntimeFilter(final RuntimeFilterWritable runtimeFilter) {
       BitData.RuntimeFilterBDef runtimeFilterDef = runtimeFilter.getRuntimeFilterBDef();
       boolean toForeman = runtimeFilterDef.getToForeman();
       QueryId queryId = runtimeFilterDef.getQueryId();
       String queryIdStr = QueryIdHelper.getQueryId(queryId);
+      runtimeFilter.retainBuffers(1);
```

Comment:
got it.

Reply:
I have tried to transfer the buffers, but it doesn't work. The runtime filter was sent by the router ,but not gain a OK ack, and not found any error information reported. The only changed code is below:
```
      String queryIdStr = QueryIdHelper.getQueryId(queryId);
      //to foreman
      if (toForeman) {
        Foreman foreman = queries.get(queryId);
        BufferAllocator bufferAllocator = foreman.getQueryContext().getAllocator();
        RuntimeFilterWritable transferred = runtimeFilter.newRuntimeFilterWritable(bufferAllocator);
        if (foreman != null) {
          executor.execute(new Runnable() {
            @Override
            public void run() {
              final Thread currentThread = Thread.currentThread();
              final String originalName = currentThread.getName();
              currentThread.setName(queryIdStr + ":foreman:routeRuntimeFilter");
              try {
                //foreman.getRuntimeFilterRouter().register(runtimeFilter);
                foreman.getRuntimeFilterRouter().register(transferred);
              } catch (Exception e) {
                logger.warn("Exception while registering the RuntimeFilter", e);
              } finally {
                currentThread.setName(originalName);
                //runtimeFilter.close();
                transferred.close();
              }
            }
          });
        }
      } else {
```
Could you give any help ?

Reply:
When you say it doesn't work what exactly is the issue with this approach ?

Reply:
The code I presented above doesn't work, as the router sends out the runtime filter without getting a Ack.OK response,no exception showed. The data tunnel which send out the runtime filter will be hanged to waiting for the ok response. This change doesn't be committed at this PR. The current committed implementation without the transferring works fine.

Reply:
I will look where the issue is with transfer code.

Reply:
I looked into the code. It's weird that you are seeing data tunnel being hanged while sending runtimeFilter since for runtime filter it doesn't look like we are using throttling listener based on semaphore count. If ACK is not received the DataTunnel shouldn't hang while sending RuntimeFilter. It will only hang while sending the DataBatches.

DataTunnel is created here for sending RuntimeFilter: https://github.com/apache/drill/pull/1504/files#diff-86eb65fdf93f77f95838f885752e660cR191
In `tunnel.sendRuntimeFilter` I don't see anywhere `semaphore.acquire` is called which will result in hang if Ack is not received. https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/rpc/data/DataTunnel.java#L94

Reply:
The router uses AccountingDataTunel which holds a SendingAccounter to send out the RF. The SendingAccounter acts as a blocking counter to book sending status.The ForemanResult’s close method will call router’s waiForComplete method which calls SendingAccounter’s synchronized method waitForSendComplete.Since without receiving Ack.Ok and failing exception ,the SendingAccounter’s batchSent counter only incremented while sending RF,no chance to decrease it ,its waiForSendComplete method will be blocked,then foreman hanged. The tricky problem here is no error message to see what happened by making this change.

Reply:
Hmm initially I didn't looked from close() call perspective. But even with respect to that it doesn't look like there will be any lost Ack case specially just with the change for transfer the buffer. It looks like more of a race condition somewhere to me. Can you open a JIRA for this issue - that we should transfer the buffers and look into about the hang issue ? So that post this PR we can look into that.